### PR TITLE
[PXT-875] performance test for openai endpoints

### PR DESCRIPTION
**Pull request by sergey-mkhitaryan**
Adds a recurring OpenAI performance test action that submits the results to Grafana.

Also includes a number of performance improvements.

### CODE REVIEW

File: pixeltable/functions/openai.py [hunk 1]
```diff
@@ -30,19 +30,60 @@
 
 _logger = logging.getLogger('pixeltable')
 
+# These are somewhat arbitrary numbers chosen to balance the throughput and the rate of errors on MacOS
```
Comment (mkornacker) [inline pixeltable/functions/openai.py]: Baking Macos/your personal environment into this file doesn't seem like a good approach.

File: pixeltable/functions/openai.py [hunk 2]
```diff
@@ -30,19 +30,60 @@
...
+MAX_KEEPALIVE_CONNECTIONS = 500
+MAX_CONNECTIONS = 2000
+TIMEOUT = 60.0
+
+
+def _adjust_fd_limit() -> None:
```
Comment (mkornacker) [inline pixeltable/functions/openai.py]: This is a general utility and not specific to OpenAI udfs. Find a place in /utils.

Also, this should probably be done during env initialization. I would expect other clients to be affected in the same way (eg, Anthropic).

File: pyproject.toml [hunk 3]
```diff
@@ -169,6 +169,7 @@ dev = [
     "azure-identity>=1.16.0",
     "twelvelabs>=1.0",
     "scenedetect>=0.6.0.1",
+    "opentelemetry-exporter-otlp>=1.38.0",
```
Comment (mkornacker) [inline pyproject.toml:174]: Why?

File: tests/test_schedulers.py [hunk 4]
```diff
@@ -6,7 +6,7 @@
 
 from pixeltable import func
 from pixeltable.exec.expr_eval.globals import Dispatcher, ExecCtx, FnCallArgs
-from pixeltable.exec.expr_eval.schedulers import RateLimitsScheduler, RequestRateScheduler
+from pixeltable.exec.expr_eval.schedulers import RateLimitsScheduler, _is_retriable_error
```
Comment (mkornacker) [inline tests/test_schedulers.py]: It's not a good idea to test the internals of modules.

File: pixeltable/exec/expr_eval/schedulers.py [hunk 5]
```diff
@@ -19,6 +19,129 @@
...
+)
+RETRIABLE_HTTP_STATUSES: dict[str, int] = {
+    'TOO_MANY_REQUESTS': HTTPStatus.TOO_MANY_REQUESTS.value,
+    'SERVICE_UNAVAILABLE': HTTPStatus.SERVICE_UNAVAILABLE.value,
+    'REQUEST_TIMEOUT': HTTPStatus.REQUEST_TIMEOUT.value,
+    'GATEWAY_TIMEOUT': HTTPStatus.GATEWAY_TIMEOUT.value,
+}
+
+
+def _is_retriable_error(exc: Exception) -> tuple[bool, float | None]:
```
Comment (mkornacker) [inline pixeltable/exec/expr_eval/schedulers.py]: These look like general http-related utilities, maybe move into a new module in /utils?

File: tests/test_schedulers.py (hunk 4 above)
Comment (sergey-mkhitaryan) [reply tests/test_schedulers.py]: do you have a specific suggestion here?

File: pyproject.toml (hunk 3 above)
Comment (sergey-mkhitaryan) [reply pyproject.toml:174]: This is used to send the perftest results to Grafana. See `report_to_grafana` in `perftest_runner.py`

File: tests/test_schedulers.py (hunk 4 above)
Comment (aaron-siegel) [reply tests/test_schedulers.py]: I'll chime in here (I've made this point before) that I think whitebox testing can sometimes be ok *as a supplement to* (not a replacement for) blackbox testing. Best to avoid it when feasible but sometimes it isn't; there are certain things whose user-visible impact is hard to measure (cache behavior is a classic example).

In this case I'd argue it's quite reasonable to have this as a separate check; it'd be difficult to engineer real use cases that reliably reproduce all the different edge cases.

File: .github/workflows/performance-tests.yml [hunk 6]
```diff
@@ -0,0 +1,59 @@
+name: Performance Tests
```
Comment (aaron-siegel) [inline .github/workflows/performance-tests.yml]: I suggest this be a job in stress-tests rather than a separate workflow.

File: pixeltable/functions/openai.py [hunk 7]
```diff
@@ -30,19 +30,60 @@
...
+    """Checks and possibly updates the file descriptor limits to accommodate the max connections.
+    Returns the max connections limit to use for the OpenAI client."""
+    try:
+        import resource
```
Comment (aaron-siegel) [inline pixeltable/functions/openai.py]: Maybe better to check if we're running Windows and fail-fast, so that we're not swallowing the error if something goes wrong with the import on other systems?

File: pixeltable/functions/openai.py (hunk 1 above)
Comment (sergey-mkhitaryan) [reply pixeltable/functions/openai.py]: Removed the comment. I'll iterate on the knobs once I'm able to run perf test somewhere other than my laptop.

File: pixeltable/functions/openai.py (hunk 2 above)
Comment (sergey-mkhitaryan) [reply pixeltable/functions/openai.py]: Moved to `utils/http.py`.

Any reason why it's a bad idea for OpenAI, Anthropic, etc. clients to update this limit lazily?

File: pixeltable/utils/http.py [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+import re
+import time
+from http import HTTPStatus
+from random import random
+from typing import Any
+
+_logger = logging.getLogger('pixeltable')
+
+
+def set_file_descriptor_limit(preferred_limit: int) -> None:
```
Comment (aaron-siegel) [inline pixeltable/utils/http.py]: Does this really belong in utils.http? Something like utils.system would be more logical.

**Review by aaron-siegel — APPROVED** [review]
